src/generators/shared.ts

```diff
@@ -0,0 +1,248 @@
+import { ISanitizer } from '@jupyterlab/apputils';
+import { Cell } from '@jupyterlab/cells';
+import { IHeading } from '../toc';
+
+const VDOM_MIME_TYPE = 'application/vdom.v1+json';
+
+const HTML_MIME_TYPE = 'text/html';
+
+export interface INotebookHeading extends IHeading {
+  numbering?: string | null;
+  type: string;
+  prompt?: string;
+  cellRef?: Cell;
+  hasChild?: boolean;
+}
+
+function incrementNumberingDict(dict: any, level: number) {
+  if (dict[level + 1] != undefined) {
+    dict[level + 1] = undefined;
+  }
+  if (dict[level] === undefined) {
+    dict[level] = 1;
+  } else {
+    dict[level]++;
+  }
+}
+
+export function generateNumbering(numberingDict: any, level: number) {
+  let numbering = undefined;
+  if (numberingDict != null) {
+    incrementNumberingDict(numberingDict, level);
+    numbering = '';
+    for (var j = 1; j <= level; j++) {
+      numbering +=
+        (numberingDict[j] == undefined ? '0' : numberingDict[j]) + '.';
+      if (j == level) {
+        numbering += ' ';
+      }
+    }
+  }
+  return numbering;
+}
+
+/**
+ * Given an HTML element, generate ToC headings
+ * by finding all the headers and making IHeading objects for them.
+ */
+export function getRenderedHTMLHeadings(
+  node: HTMLElement,
+  onClickFactory: (el: Element) => (() => void),
+  sanitizer: ISanitizer,
+  numberingDict: any,
+  lastLevel: number,
+  needNumbering = false,
+  cellRef?: Cell
+): INotebookHeading[] {
+  let headings: INotebookHeading[] = [];
+  let headingNodes = node.querySelectorAll('h1, h2, h3, h4, h5, h6, p');
+  if (headingNodes.length > 0) {
+    let markdownCell = headingNodes[0];
```

Comment:
I see here is where you are only grabbing the first header (or `p` tag?). Why are you naming this `markdownCell`. Shouldn't this also apply to rendered markdown documents, or is there something I am missing?

Reply:
Because we are only considering the first line for markdown cells but we need to consider the whole rendered markdown document, I use a different function to render the headings for markdown documents.

Reply:
Okay, sounds good.

Reply:
What do you think about headers in outputs?  Should we also only consider the first one?

Reply:
In the outputs right now we only considers the first one, but we can change it very quick. I'm also a little bit concerned about the show/hide feature for these outputs but it is really something in the future